src/Chapter3.hs

```diff
@@ -1005,7 +1131,26 @@ properties using typeclasses, but they are different data types in the end.
 Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
-
+newtype Health' = Health' { getHealth' :: Int }
+newtype Attack' = Attack' { getAttack' :: Int }
+newtype Defense' = Defense' { getDefense' :: Int }
+
+data Monster' = Monster'
+  { monsterHealth' :: Health'
+  , monsterAttack' :: Attack'
+  }
+
+data Knight' = Knight'
+  { knightHealth' :: Health'
+  , knightAttack' :: Attack'
+  , knightDefense' :: Defense'
+  }
+
+class Fighter a where
+  fight' :: a -> b -> String -> (a, b)
+
+doFight :: (Fighter a, Fighter b) => a -> b -> (a, b)
+doFight fighter1 fighter2 = undefined
```

Comment:
Yes, the error message might not be readable. But what is means is that type variable `b` in the definition of the `fight'` method literally can be anything. So, a valid instance of `Figther` for `Knight'` will look like this:

```haskell
instance Fighter Knight' where
    fight' :: Knight' -> b -> String -> (Knight', b)
```

You see that you have no information about `b` type variable. It can be anything! `Char` or `Bool` or list. You can't fix this type during instance implementation. That's why nothing connects `b` with values the `doFight` function.

You can solve this problem in a multiple ways, but what you can explore is adding a `Fighter` constraint to the `fight'` method:

```haskell
class Fighter a where
  fight' :: Fighter b => a -> b -> String -> (a, b)
```

After that, your instance will look like this:

```haskell
instance Fighter Knight' where
    fight' :: Fighter b => Knight' -> b -> String -> (Knight', b)
```

Now, when implementing this instance, you have access to all the internal implementation details of a `Knight'`, but you also know that type `b` implements the `Fighter` typeclass. You can add more methods to `Fighther` to work with `b` in a generic way.